prov/shm/src/smr_init.c

```diff
@@ -91,6 +133,10 @@ static int smr_getinfo(uint32_t version, const char *node, const char *service,
 	if (ret)
 		return ret;
 
+	ret = smr_shm_space_check((*info)->tx_attr->size, (*info)->rx_attr->size);
```

Comment:
Is this the right place to put this? If one peer allocates multiple endpoints will this still be an accurate?

Reply:
I think so, because we are not modifying the tx/rx_attr->size after util_getinfo call. And we can calculate the upper limit of shm space usage we need. Then it returns the eligible fi_info  to either apps or in conjunction used other provider based on the space check result.
 
If I understand correctly, the SMR_MAX_PEERS means the maximum local endpoints, even though we named it as it is now. Each endpoint has an address and will be inserted into the smr_map.
```
struct smr_map {
        fastlock_t      lock;
        struct smr_peer peers[SMR_MAX_PEERS];
};
```
And the  maximum local endpoint cannot be large than SMR_MAX_PEERS.

Reply:
MAX_PEERS is the number of remote endpoints that a single local endpoint can communicate with.  There's no limit on the number of local endpoints that can be created, as long as we have enough memory.  It's highly likely that an app will only allocate a single local endpoint for shared memory communication.  So this likely works as a first pass sanity check.  We should simply fail the endpoint creation later if the app actually tries to go beyond the limit.

Note that other providers do not fail getinfo calls based on dynamically checking if a new endpoint _could_ be created based on current resource allocations.  I'm not sure we want that.  So, I think it's reasonable to have a check that validates that the request is reasonable for the _system_ configuration, and not worry about failing getinfo because the app has too many endpoints open.  That failure can come when trying to open the endpoint.

Reply:
I see your point. 
But in shm provider specifically, we have this SMR_MAX_PEERS hard limit (256 currently), I think it's safe to assume that an app can have at most SMR_MAX_PEERS local endpoints, because shm av cannot handle more than SMR_MAX_PEERS addresses. 

In addition, I'm assuming you meant having the check per endpoint during ep_open(), e.g., when ep0 dose ep_open(), and it checks if the shm size it needs is less than the current available shm space. If it's not, the end_open fails. 
There's a potential issue here. Some sort of lock/synchronization is needed across all local endpoints, when they try to calculate the current available shm space. Do you have any idea/thought on this?

Reply:
We're making a assumptions here about how an app allocates its endpoints and is communicating with its peers.  It's confusing for the code to use a value defined as one thing as a stand in replacement for another.

Regardless, it seems that the goal here is to validate that the requested fi_getinfo call can be supported based on current system configuration.  I'm suggesting that we don't even try to track the actual usage and instead just validate the fi_getinfo input.  At a minimum we can assume that there is at least 1 peer that we need to communicate with, which would multiply the result by 2.  For a maximum test value, I would use the number of CPUs as the most likely value.

Adding this check has the possibility of failing fi_getinfo calls that are currently passing, and breaking currently working apps.  That should not happen if we use the minimum check of times 2, but may if we use x #CPUs.  Using x 256 increases the chance of breaking an app even more.